holoviews/operation/datashader.py

```diff
@@ -20,75 +22,23 @@
+from ..core.sheetcoords import BoundingBox
 from ..core.util import get_param_values, basestring
 from ..element import Image, Path, Curve, Contours, RGB
 from ..streams import RangeXY, PlotSize
 from ..plotting.util import fire
 
-@dispatch(Element)
-def discover(dataset):
-    """
-    Allows datashader to correctly discover the dtypes of the data
-    in a holoviews Element.
-    """
-    return dsdiscover(PandasInterface.as_dframe(dataset))
-
-
-@bypixel.pipeline.register(Element)
-def dataset_pipeline(dataset, schema, canvas, glyph, summary):
-    """
-    Defines how to apply a datashader pipeline to a holoviews Element,
-    using multidispatch. Returns an Image type with the appropriate
-    bounds and dimensions. Passing the returned Image to datashader
-    transfer functions is not yet supported.
-    """
-    x0, x1 = canvas.x_range
-    y0, y1 = canvas.y_range
-    kdims = [dataset.get_dimension(d) for d in (glyph.x, glyph.y)]
-
-    if dataset.interface is PandasInterface:
-        agg = pandas_pipeline(dataset.data, schema, canvas,
-                              glyph, summary)
-    elif dataset.interface is DaskInterface:
-        agg = dask_pipeline(dataset.data, schema, canvas,
-                            glyph, summary)
-
-    agg = agg.rename({'x_axis': kdims[0].name,
-                      'y_axis': kdims[1].name})
-    return agg
 
-
-class aggregate(Operation):
-    """
-    aggregate implements 2D binning for any valid HoloViews Element
-    type using datashader. I.e., this operation turns a HoloViews
-    Element or overlay of Elements into an hv.Image or an overlay of
-    hv.Images by rasterizing it, which provides a fixed-sized
-    representation independent of the original dataset size.
-
-    By default it will simply count the number of values in each bin
-    but other aggregators can be supplied implementing mean, max, min
-    and other reduction operations.
-
-    The bins of the aggregate are defined by the width and height and
-    the x_range and y_range. If x_sampling or y_sampling are supplied
-    the operation will ensure that a bin is no smaller than the minimum
-    sampling distance by reducing the width and height when the zoomed
-    in beyond the minimum sampling distance.
-
-    By default, the PlotSize stream is applied when this operation
-    is used dynamically, which means that the height and width
-    will automatically be set to match the inner dimensions of 
-    the linked plot.
-    """
-
-    aggregator = param.ClassSelector(class_=ds.reductions.Reduction,
-                                     default=ds.count())
+class resample_operation(Operation):
 
     dynamic = param.Boolean(default=True, doc="""
        Enables dynamic processing by default.""")
 
+    expand = param.Boolean(default=True, doc="""
+       Whether the x_range and y_range should be allowed to expand
+       beyond the extent of the data.""")
```

Comment:
Actually there's a good reason not to enable it for ``aggregate``. Computing the actual ``range`` is expensive when you have to iterate over all the datapoints and it doesn't actually gain you anything except making sure that pixels aren't wasted on regions where there is no data. I could set it to ``True`` for regrid only, but it might be better to be consistent.